README.md
# MutableDataScanner

[![Build Status](https://www.bitrise.io/app/b6bf3a2353584846.svg?token=AfNOSpFuWfABDb-o68wScg&branch=master)](https://www.bitrise.io/app/b6bf3a2353584846)
[![Carthage compatible](https://img.shields.io/badge/Carthage-compatible-4BC51D.svg?style=flat)](https://github.com/Carthage/Carthage)
[![](https://img.shields.io/badge/Xcode-7.3%2B-brightgreen.svg?style=flat)]()
[![](https://img.shields.io/badge/iOS-8.0%2B-brightgreen.svg?style=flat)]()
[![](https://img.shields.io/badge/OS%20X-10.10%2B-brightgreen.svg?style=flat)]()

A simple text scanner which can parse NSMutableData using delimiter.

Faster because it does not have to do a NSData <-> String conversion.

It can be easily and reliably parse of the Twitter Streaming API and other stream.


### Performance

Test: https://github.com/s-aska/MutableDataScanner/blob/master/MutableDataScannerTests/MutableDataScannerTests.swift

- MutableDataScanner specific delimiter ... average: __0.071__
- MutableDataScanner line delimiter ( CRLF or LF ) ... average: 0.092
- String#componentsSeparatedByString ... average: 0.304


## Usage

### for Twitter Streaming API

See [Processing streaming data](https://dev.twitter.com/streaming/overview/processing) for information about the Parsing responses you will receive from the streaming API.

```swift
import Foundation
import MutableDataScanner
import SwiftyJSON

class TwitterAPIStreamingRequest: NSObject, NSURLSessionDataDelegate {

    let scanner = MutableDataScanner(delimiter: "\r\n")

    // ...

    func connection(connection: NSURLConnection, didReceiveData data: NSData) {
        self.scanner.appendData(data)
        while let data = self.scanner.next() {
            if data.length > 0 {
                let json = JSON(data: data)
            }
        }
    }
}
```

### Bytes

- `read(length: Int) -> NSData?`

```swift
let data = "0123456789abcdefghijklmnopqrstuvwxyz".dataUsingEncoding(NSUTF8StringEncoding)!
let scanner = MutableDataScanner(data: data)

scanner.read(length: 10) // => 0123456789
scanner.read(length: 10) // => abcdefghij
scanner.read(length: 10) // => klmnopqrst
scanner.read(length: 10) // => uvwxyz
scanner.read(length: 10) // => (nil)
```

- `read(offset: Int, length: Int) -> NSData?`

```swift
let data = "0123456789abcdefghijklmnopqrstuvwxyz".dataUsingEncoding(NSUTF8StringEncoding)!
let scanner = MutableDataScanner(data: data)

scanner.read(offset: 5, length: 5) // => 56789
scanner.read(offset: 5, length: 5) // => fghij
scanner.read(offset: 5, length: 5) // => pqrst
scanner.read(offset: 5, length: 5) // => z
scanner.read(offset: 5, length: 5) // => (nil)
```

### Delimiter

#### Specify the delimiter in the properties.

- `init(data: NSMutableData = default, delimiter: NSData)`
- `init(data: NSMutableData = default, delimiter: String)`
- `next() -> NSData?`
- `hasNext() -> Bool`

```swift
let data = "0123456789\nabcdefghijklmnopqrstuvwxyz\n0123".dataUsingEncoding(NSUTF8StringEncoding)!
let scanner = MutableDataScanner(data: data, delimiter: "\n")

scanner.data // => 0123456789\nabcdefghijklmnopqrstuvwxyz\n0123

scanner.hasNext() // => true
scanner.next() // => 0123456789
scanner.data // => abcdefghijklmnopqrstuvwxyz\n0123

scanner.hasNext() // => true
scanner.next() // => abcdefghijklmnopqrstuvwxyz
scanner.data // => 0123

scanner.hasNext() // => false
scanner.next() // => (nil)
```

#### Specify the delimiter in the arguments.

- `next(delimiter: NSData) -> NSData?`
- `hasNext(delimiter: NSData) -> Bool`
- `next(delimiter: String) -> NSData?`
- `hasNext(delimiter: String) -> Bool`

```swift
let data = "0123456789\nabcdefghijklmnopqrstuvwxyz\n0123".dataUsingEncoding(NSUTF8StringEncoding)!
let scanner = MutableDataScanner(data: data)

scanner.data // => 0123456789\nabcdefghijklmnopqrstuvwxyz\n0123

scanner.hasNext("\r\n") // => false
scanner.hasNext("\n") // => true
scanner.next("\r\n") // => (nil)
scanner.next("\n") // => 0123456789
scanner.data // => abcdefghijklmnopqrstuvwxyz\n0123
```

#### CRLF or LF

- `nextLine() -> NSData?`
- `hasNextLine -> Bool`

```swift
let data = "0123456789\r\nabcdefghijklmnopqrstuvwxyz\n0123".dataUsingEncoding(NSUTF8StringEncoding)!
let scanner = MutableDataScanner(data: data)

scanner.data // => 0123456789\r\nabcdefghijklmnopqrstuvwxyz\n0123

scanner.hasNextLine() // => true
scanner.nextLine() // => 0123456789
scanner.data // => abcdefghijklmnopqrstuvwxyz\n0123

scanner.hasNextLine() // => true
scanner.nextLine() // => abcdefghijklmnopqrstuvwxyz
scanner.data // => 0123

scanner.hasNextLine() // => false
scanner.nextLine() // => (nil)
```


## Requirements

- iOS 8.0+ / Mac OS X 10.10+
- Xcode 7.3+


## Installation

#### Carthage

Add the following line to your [Cartfile](https://github.com/carthage/carthage)

```
github "s-aska/MutableDataScanner"
```

#### CocoaPods

Add the following line to your [Podfile](https://guides.cocoapods.org/)

```
use_frameworks!
pod 'MutableDataScanner'
```


## License

MutableDataScanner is released under the MIT license. See LICENSE for details.


CHANGELOG.md
# 1.0.8

Swift 2.2　optimization
SwiftLint

# 1.0.7

Rename scheme name

# 1.0.0

Stable Release

- [x] `appendData(NSData)`
- [x] `read(length: Int)`
- [x] `read(offset: Int, length: Int)`
- [x] `next()`
- [x] `next(delimiter: NSData)`
- [x] `next(delimiter: String)`
- [x] `hasNext()`
- [x] `hasNext(delimiter: NSData)`
- [x] `hasNext(delimiter: String)`
- [x] `nextLine()`
- [x] `hasNextLine()`


MutableDataScanner/MutableDataScanner.swift
//
//  MutableDataScanner.swift
//  MutableDataScanner
//
//  Created by Shinichiro Aska on 8/20/15.
//  Copyright © 2015 Shinichiro Aska. All rights reserved.
//

import Foundation

/// A simple text scanner which can parse NSMutableData using delimiter module, but as a class.
public class MutableDataScanner {

    /// Constants to use nextLine() and hasNextLine().
    struct Static {
        static let dataCR = "\r".dataUsingEncoding(NSUTF8StringEncoding)!
        static let dataLF = "\n".dataUsingEncoding(NSUTF8StringEncoding)!
    }

    /// buffer.
    public var data: NSMutableData

    /// delimiter to use with no arguments next() and hasNext().
    public var delimiter: NSData?

    /**
    Create A MutableDataScanner Instance

    - parameter data: buffer.
    */
    public init(data: NSMutableData = NSMutableData()) {
        self.data = data
        self.delimiter = nil
    }

    /**
    Create A MutableDataScanner Instance

    - parameter data: buffer.
    - parameter delimiter: to use with no arguments next() and hasNext().
    */
    public init(data: NSMutableData = NSMutableData(), delimiter: NSData) {
        self.data = data
        self.delimiter = delimiter
    }

    /**
    Create A MutableDataScanner Instance

    - parameter data: buffer.
    - parameter delimiter: to use with no arguments next() and hasNext().
    */
    public init(data: NSMutableData = NSMutableData(), delimiter: String) {
        self.data = data
        self.delimiter = delimiter.dataUsingEncoding(NSUTF8StringEncoding)!
    }

    /**
    Appends the content of another NSData object to the buffer.
    The data object whose content is to be appended to the contents of the buffer.

    - parameter data: The data object whose content is to be appended to the contents of the buffer.
    */
    public func appendData(data: NSData) {
        self.data.appendData(data)
    }

    /**
    It returns data for the specified length from the specified read start position,
     and then removed from the buffer.

    - parameter offset: reading start position
    - parameter length: reading data length

    - returns: Data of specified length
    */
    public func read(offset offset: Int, length: Int) -> NSData? {
        if offset > data.length {
            return nil
        }
        let length = min(length, data.length - offset)
        let chunk = data.subdataWithRange(NSRange.init(location: offset, length: length))
        data.forward(offset + length)
        return chunk
    }

    /**
    It returns data for the specified length, and then removed from the buffer.

    - parameter length: reading data length

    - returns: Data of specified length
    */
    public func read(length length: Int) -> NSData? {
        if data.length == 0 {
            return nil
        }
        let length = min(length, data.length)
        let line = data.subdataWithRange(NSRange.init(location: 0, length: length))
        data.forward(length)
        return line
    }

    /**
    Returns true if it contains a delimiter in buffer.

    - returns: true if it contains a delimiter in buffer.
    */
    public func hasNext() -> Bool {
        guard let delimiter = delimiter else {
            fatalError("hasNext() need delimiter."
                + " eg: MutableDataScanner(delimiter: NSData or String)")
        }
        return self.hasNext(delimiter)
    }

    /**
    It returns the data to the next delimiter, and removes it from the buffer.
    If there is no delimiter in the buffer, it returns nil.
    It does not include delimiter in the data.

    - returns: data to the next delimiter.
    */
    public func next() -> NSData? {
        guard let delimiter = delimiter else {
            fatalError("next() need delimiter. eg: MutableDataScanner(delimiter: NSData or String)")
        }
        return self.next(delimiter)
    }

    /**
    Returns true if it contains a delimiter in buffer.

    - parameter delimiter: delimiter data

    - returns: true if it contains a delimiter in buffer.
    */
    public func hasNext(delimiter: String) -> Bool {
        guard let delimiter = delimiter.dataUsingEncoding(NSUTF8StringEncoding) else {
            fatalError("dataUsingEncoding(NSUTF8StringEncoding) failure.")
        }
        return self.hasNext(delimiter)
    }

    /**
    It returns the data to the next delimiter, and removes it from the buffer.
    If there is no delimiter in the buffer, it returns nil.
    It does not include delimiter in the data.

    - parameter delimiter: delimiter data

    - returns: data to the next delimiter.
    */
    public func next(delimiter: String) -> NSData? {
        guard let delimiter = delimiter.dataUsingEncoding(NSUTF8StringEncoding) else {
            fatalError("dataUsingEncoding(NSUTF8StringEncoding) failure.")
        }
        return self.next(delimiter)
    }

    /**
    Returns true if it contains a delimiter in buffer.

    - parameter delimiter: delimiter data

    - returns: true if it contains a delimiter in buffer.
    */
    public func hasNext(delimiter: NSData) -> Bool {
        let range = data.rangeOfData(delimiter)
        return range.location != NSNotFound
    }

    /**
    It returns the data to the next delimiter, and removes it from the buffer.
    If there is no delimiter in the buffer, it returns nil.
    It does not include delimiter in the data.

    - parameter delimiter: delimiter data

    - returns: data to the next delimiter.
    */
    public func next(delimiter: NSData) -> NSData? {
        let range = data.rangeOfData(delimiter)
        if range.location != NSNotFound {
            let line = data.subdataWithRange(NSRange.init(location: 0, length: range.location))
            data.forward(range.location + range.length)
            return line
        } else {
            return nil
        }
    }

    /**
    Returns true if the buffer there is a line break
    It considers the CRLF or LF and line feed.

    - returns: true if the buffer there is a line break
    */
    public func hasNextLine() -> Bool {
        let range = data.rangeOfData(Static.dataLF)
        return range.location != NSNotFound
    }

    /**
    It returns the following line, and then removed from the buffer.
    If there is no new line in the buffer, it returns nil.
    It considers the CRLF or LF and line feed.

    - returns: the next line
    */
    public func nextLine() -> NSData? {
        let range = data.rangeOfData(Static.dataLF)
        if range.location != NSNotFound {
            let line: NSData
            let end = NSRange.init(location: range.location - 1, length: 1)
            if data.subdataWithRange(end) == Static.dataCR {
                line = data.subdataWithRange(NSRange.init(location: 0, length: range.location - 1))
            } else {
                line = data.subdataWithRange(NSRange.init(location: 0, length: range.location))
            }
            data.forward(range.location + range.length)
            return line
        } else {
            return nil
        }
    }
}

private extension NSMutableData {
    func forward(length: Int) {
        let range = NSRange.init(location: 0, length: length)
        self.replaceBytesInRange(range, withBytes: nil, length: 0)
    }
}

private extension NSData {
    func rangeOfData(dataToFind: NSData) -> NSRange {
        return self.rangeOfData(dataToFind,
                                options: NSDataSearchOptions(rawValue: 0),
                                range: NSRange.init(location: 0, length: self.length))
    }
}


MutableDataScannerTests/MutableDataScannerTests.swift
//
//  MutableDataScannerTests.swift
//  MutableDataScannerTests
//
//  Created by Shinichiro Aska on 8/20/15.
//  Copyright © 2015 Shinichiro Aska. All rights reserved.
//

import XCTest
import MutableDataScanner

class MutableDataScannerTests: XCTestCase {

    override func setUp() {
        super.setUp()
    }

    override func tearDown() {
        super.tearDown()
    }

    func testInitDefault() {
        let scanner = MutableDataScanner()
        XCTAssertEqual(scanner.delimiter, nil)
    }

    func testInitWithData() {
        let data = NSMutableData()
        let scanner = MutableDataScanner(data: data)
        data.appendData("012345".dataValue)
        XCTAssertEqual(scanner.data, data)
    }

    func testInitWithDataAndDelimiterString() {
        let data = NSMutableData()
        let scanner = MutableDataScanner(data: data, delimiter: "\t")
        data.appendData("012345".dataValue)
        XCTAssertEqual(scanner.data, data)
        XCTAssertEqual(scanner.delimiter?.stringValue, "\t")
    }

    func testInitWithDelimiterData() {
        let scanner = MutableDataScanner(delimiter: "\t".dataValue)
        XCTAssertEqual(scanner.delimiter?.stringValue, "\t")
    }

    func testInitWithDelimiterString() {
        let scanner = MutableDataScanner(delimiter: "\t")
        XCTAssertEqual(scanner.delimiter?.stringValue, "\t")
    }

    func testReadLength() {
        let scanner = MutableDataScanner()
        let data = "0123456789abcdefghijklmnopqrstuvwxyz".dataValue
        scanner.appendData(data)
        XCTAssertEqual(scanner.read(length: 10)!.stringValue, "0123456789", "read length")
        XCTAssertEqual(scanner.read(length: 10)!.stringValue, "abcdefghij", "read length")
        XCTAssertEqual(scanner.read(length: scanner.data.length)!.stringValue,
                       "klmnopqrstuvwxyz", "read length")
        XCTAssertEqual(scanner.read(length: 1), nil)
    }

    func testReadLengthOver() {
        let scanner = MutableDataScanner()
        let data = "012345".dataValue
        scanner.appendData(data)
        XCTAssertEqual(scanner.read(length: 100)!.stringValue, "012345", "read length")
        XCTAssertEqual(scanner.read(length: 1), nil)
    }

    func testReadOffsetLength() {
        let scanner = MutableDataScanner()
        let data = "0123456789abcdefghijklmnopqrstuvwxyz".dataValue
        scanner.appendData(data)
        XCTAssertEqual(scanner.read(offset: 3, length: 7)!.stringValue, "3456789", "read length")
        XCTAssertEqual(scanner.read(offset: 3, length: 7)!.stringValue, "defghij", "read length")
        XCTAssertEqual(scanner.read(offset: 3, length: scanner.data.length)!.stringValue,
                       "nopqrstuvwxyz", "read length")
        XCTAssertEqual(scanner.read(offset: 3, length: 1), nil)
    }

    func testNextLine() {
        let scanner = MutableDataScanner()
        let data = "1\n1\r\n1\r1".dataValue
        var count = 0
        scanner.appendData(data)
        while let line = scanner.nextLine() {
            XCTAssertEqual(line.stringValue, "1")
            count += 1
        }
        XCTAssertEqual(count, 2, "data count")
        XCTAssertEqual(scanner.data.length, 3, "buffer length")
    }

    func testNext() {
        let scanner = MutableDataScanner(delimiter: "\r\n")
        let data = "012345\nabcdefg\r\n".dataValue
        var count = 0
        scanner.appendData(data)
        while let line = scanner.next() {
            NSLog("\(line.stringValue)")
            assert(line.stringValue == "012345\nabcdefg")
            count += 1
        }
        XCTAssertEqual(count, 1, "data count")
        XCTAssertEqual(scanner.data.length, 0, "buffer length")
    }

    func testPerformanceAutoDelimiter() {
        self.measureBlock {
            let scanner = MutableDataScanner()
            let data = "12345\n12345\r\n12345\r12345\n12345\r\n12345".dataValue
            for _ in 1...10000 {
                scanner.appendData(data)
                while let _ = scanner.nextLine() {
                }
            }
        }
    }

    func testPerformanceSpecificDelimiter() {
        self.measureBlock {
            let scanner = MutableDataScanner(delimiter: "\n")
            let data = "12345\n12345\r\n12345\r12345\n12345\r\n12345".dataValue
            for _ in 1...10000 {
                scanner.appendData(data)
                while let _ = scanner.next() {
                }
            }
        }
    }

    func testPerformanceSplitReader() {
        let data = "12345\n12345\r\n12345\r12345\n12345\r\n12345".dataValue
        let buffer = NSMutableData()
        self.measureBlock {
            for _ in 1...10000 {
                buffer.appendData(data)
                if let string = NSString(data: buffer, encoding: NSUTF8StringEncoding) {
                    var array = string.componentsSeparatedByString("\n")
                    if let last = array.popLast()?.dataUsingEncoding(NSUTF8StringEncoding)! {
                        buffer.setData(last)
                    } else {
                        buffer.setData(NSData())
                    }
                    for _ in array {
                    }
                }
            }
        }
    }
}

private extension NSData {
    var stringValue: NSString {
        return NSString(data: self, encoding: NSUTF8StringEncoding)!
    }
}


private extension String {
    var dataValue: NSData {
        return self.dataUsingEncoding(NSUTF8StringEncoding)!
    }
}
